### DISCUSSION TITLE
WGSL bitwise operator error: shift left operation results in sign change

### DISCUSSION THREAD

**Discussion by coramat** [gpuweb/gpuweb#4169] [Q&A] [answered]
I have this snippet of code that I translated from hlsl (properly working):
```
const pY: i32 = 1136930381;
let var2 : i32 = (pY << 1);
```
but in wgsl I am getting this error:
```
error: shift left operation results in sign change
runtime.js:8 let var2 : i32 = PRIME_Y << 1 ;
```

Does anyone know what is the problem here and if there is a workaround that does not modify the output?

**Comment by coramat** [comment]
I do not have any issues if I do the same operation with pX instead of pY:
```
const pX: i32 = 501125321;
let var1 : i32 = (pX << 1);
let var3 : i32 = (-pX << 1);
```

**Comment by ben-clayton** [comment, marked as answer]
`1136930381` is `0x43C42E4D` in hex. The most significant bit of an `i32` is the sign bit, and left shifting `0x43C42E4D` by one would result in the number changing sign.

[The WGSL spec says:](https://www.w3.org/TR/WGSL/#bit-expr)

> When both e1 and e2 are known before shader execution start, the result must not overflow:
>If T is a signed integer type, and the e2+1 most significant bits of e1 do not have the same bit value, then:
> It is a shader-creation error if e1 and e2 are const-expressions.

There are two workarounds:
1. Drop the explicit `i32`  type from the `const` declaration of `pY`:

```
const pY = 1136930381;
let var2 : i32 = (pY << 1);
```

This will give `pY` the type of [AbstractInt](https://www.w3.org/TR/WGSL/#abstract-types), which has 64 bits of precision. Note: you'll still get an error if the initalizer expression of `var2` does not fit in an `i32`.

2. Cast to `u32` before shifting, and then cast back to `i32`: `i32(u32(pY) << 1)`. You're then being explicit about allowing a change in sign. Again, you'll error if you push the bits outside of the `u32`.

I hope this helps.

---

Update: on second thought, the first suggestion won't work, as the left-shifted number doesn't fit in an `i32`. If it did, you wouldn't have got the error in the first place!

**Comment by coramat** [comment]
Thanks a lot for the fast reply! 
I went with the second option before posting this issue but I was not sure. 
The explanation makes a lot of sense, thanks again :)
